torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1820,6 +1831,896 @@ def clip_grad_norm_(
+            flat_param_state (Dict[str, Any]): Entry for a flattened parameter
+                in the "state" part of the optimizer state dict.
+            fsdp_module (FullyShardedDataParallel): FSDP module that owns the
+                flattened parameter, i.e. holds it in ``self.params``.
+
+        Returns:
+            unflat_param_state (Dict[int, Dict[str, Any]]): A :class:`dict`
+                holding the entries in the "state" part of the optimizer state
+                dict corresponding to the unflattened parameters comprising the
+                flattened parameter ``flat_param``. The keys are indexed
+                starting from zero and need to all be increased by an offset
+                for the overall optimizer state dict.
+        """
+        assert sum(p is flat_param for p in fsdp_module.params) == 1, \
+            "`fsdp_module` must own `flat_param`"
+        tensor_state, non_tensor_state, zero_dim_tensor_state, pad_per_rank = \
+            fsdp_module._communicate_optim_state(
+                flat_param, flat_param_state, target_rank,
+            )
+        unflat_param_state = fsdp_module._unflatten_communicated_optim_state(
+            flat_param,
+            tensor_state,
+            non_tensor_state,
+            zero_dim_tensor_state,
+            pad_per_rank,
+        )
+        return unflat_param_state
+
+    def _communicate_optim_state(
+        self,
+        flat_param: FlatParameter,
+        flat_param_state: Dict[str, Any],
+        target_rank: int = 0,
+    ) -> Tuple[Dict[str, List], Dict[str, List], Dict[str, List], List[int]]:
+        """
+        Communicates the optimizer state for a flattened parameter
+        ``flat_param`` across ranks so that the target rank ``target_rank``
+        holds the entire non-sharded optimizer state.
+
+        If ``N`` is the number of tensor optimizer states in the optimizer
+        state dict, then the communication complexity is 0 if ``N = 0`` and
+        ``N + 1`` otherwise (where the plus 1 comes from all-gathering the
+        padding per rank).
+
+        Args:
+            flat_param (FlatParameter): The flattened parameter.
+            flat_param_state (Dict[str, Any]): The entry in the "state" part of
+            the optimizer state dict corresponding to the flattened parameter.
+            target_rank (int): Rank on which to consolidate the optimizer
+                state. (Default: 0)
+
+        Returns:
+            tensor_state (Dict[str, List]): Mapping from positive-dimension
+                tensor state names to :class:`list` s of positive-dimension
+                tensor state values with length equal to the world size.
+            non_tensor_state (Dict[str, List]): Mapping from non-tensor state
+                names to :class:`list` s of non-tensor state values with length
+                equal to the world size.
+            zero_dim_tensor_state (Dict[str, List]): Mapping from zero-
```

Comment:
Seems like this is always an empty dictionary? If so, why do we have this at all?